packages/vitest/src/api/setup.ts

```diff
@@ -39,6 +39,8 @@ export function setup(ctx: Vitest) {
         },
         async onFinished() {
           await ctx.report('onFinished')
+          if (!ctx.config.watch && ctx.config.browser)
```

Comment:
Instead of this we should add a method to exit and call it via websockets from the browser

Reply:
I'm testing it on local: moving that logic to ws close event, adding a window `unload` listener to close the ws

Reply:
@sheremet-va added `onBrowserClosed` and works, added `onPathsCollected` to `WebSocketReporter`, re-run button clicks on ui now also working (we need to fix the paths and the cache)

Reply:
I'm doing some magic now to always use webdriverio to open browser, will remove some code there.

Reply:
here my local changes (sent also DM on discord)

[chore__allow_close_server_closing_browser_ui.patch](https://github.com/vitest-dev/vitest/files/11009449/chore__allow_close_server_closing_browser_ui.patch)